src/main/java/org/mastodon/revised/model/feature/Feature.java

```diff
@@ -0,0 +1,129 @@
+ * <p>
+ * Features management in application requires them to declare a String key that
+ * must be unique within the application scope.
+ * <p>
+ * A Feature is defined for a specific target in the model (vertex, edge, graph,
+ * etc.) which must be specified. The feature values for each item of the model
+ * are stored in a {@link PropertyMap} that can be accessed.
+ * <p>
+ * Since many visualization tools in Mastodon deal with scalar, real feature
+ * values, the computer must also be able to return a set of feature
+ * projections, that translate or project the feature on several real scalar
+ * values. Since the computer defines what the feature it compute is, it also
+ * must be able to define the meaningful projections for the feature.
+ * 
+ * @author Jean-Yves Tinevez
+ *
+ * @param <O>
+ *            the type of the object this feature is defined for (target).
+ * @param <T>
+ *            the type of the values of this feature.
+ * @param <K>
+ *            the type of the property map used to store feature values.
+ */
+public class Feature< O, T, K extends PropertyMap< O, T > >
+{
+
+	/**
+	 * The feature unique key.
+	 */
+	private final String key;
+
+	/**
+	 * The feature property map.
+	 */
+	private final K propertyMap;
+
+	/**
+	 * The feature projections, stored as a map from projection names to
+	 * projections.
+	 */
+	private final Map< String, FeatureProjection< O > > projections;
+
+	/**
+	 * The feature target.
+	 */
+	private final FeatureTarget target;
+
+	/**
+	 * Creates a new immutable feature instance.
+	 * 
+	 * @param key
+	 *            The feature unique key. Must be unique within the application
+	 *            scope.
+	 * @param target
+	 *            The feature target.
+	 * @param propertyMap
+	 *            The feature property map.
+	 * @param projections
+	 *            The feature projections, stored as a map from projection names
+	 *            to projections.
+	 */
+	public Feature( final String key, final FeatureTarget target, final K propertyMap, final Map< String, FeatureProjection< O > > projections )
+	{
+		this.key = key;
+		this.target = target;
+		this.propertyMap = propertyMap;
+		this.projections = projections;
+	}
+
+	/**
+	 * Returns the key of this feature.
+	 * 
+	 * @return the feature key.
+	 */
+	public String getKey()
+	{
+		return key;
+	}
+
+	/**
+	 * Returns the property map storing the values of this feature.
+	 * 
+	 * @return the property map.
+	 */
+	public K getPropertyMap()
+	{
+		return propertyMap;
+	}
+
+	/**
+	 * Returns the feature projections for this feature.
+	 * <p>
+	 * They are returned as a map from projection keys to actual projections.
+	 * 
+	 * @return the map of feature projections.
+	 */
+	public Map< String, FeatureProjection< O > > getProjections()
+	{
+		return projections;
+	}
+
+	/**
+	 * Returns the object this feature is defined on, as a
+	 * {@link FeatureTarget}.
+	 * 
+	 * @return the feature target.
+	 */
+	public FeatureTarget getTarget()
```

Comment:
Why is there a `FeatureTarget` enum when there is also a generic parameter `O` that defines the target?

Reply:
I could not come with one. 
After type erasure, the class does not expose anything that is `O`. So I deferred to explicit declaration with the enum.

Reply:
You could pass `Class<O>` as a constructor param instead of `FeatureTarget`?

Reply:
Noted! Thanks!